File: document/js_wp/QuickSort.js

// 快速排序
function quickSort(a) {
    if (!a || a.length ==0) {
        return;
    }
    // 优化：打乱 a 的顺序
    shuffle(a)
    sort(a, 0, a.length-1);
}

// 打乱 a 的顺序， 避免极端的情况
function shuffle(a) {
    for (let i=0; i<a.length; i++) {
        let j = Math.floor(Math.random() * (a.length-i))
        exch(a, i, i+j);
    }
}

function sort(a, lo, hi) {
    if (lo>=hi) return;
    // 分成两半，左边的都小于a[p], 右边的都大于a[p]
    let p = partition(a, lo, hi);

    sort(a, lo, p-1);
    sort(a, p+1, hi);
}

function partition(a, lo, hi) {
    let left = lo;
    let right = hi+1;
    let j = lo;
    while (left<right) {
        while (less(a, ++left, j)) if (left==hi) break;
        while (less(a, j, --right)) if (right==lo) break;
        if (left>=right) break;
        exch(a, left, right)
    }
    exch(a, j, right);
    return right;
}

function less(a, i, j) {
    if (a[i]<a[j]) {
        return true;
    } else {
        return false;
    }
}

function exch(a, i, j) {
    let t = a[i];
    a[i] = a[j];
    a[j] = t;
}

let a = [3, 6, 1, 2, 9, 0, 1, 2, 3]
console.log(a);
quickSort(a);
console.log(a);

File: document/LeetCodeWP/src/main/java/com/hef/week09/sort/QuickSort.java
package com.hef.week09.sort;

import java.util.*;

/**
 * @Date 2021/7/1
 * @Author lifei
 */
public class QuickSort {

    public static void main(String[] args) {
        Integer[] a = {9, 1, 4, 2, 8, 1, 3, 0, 2};
        System.out.println(Arrays.toString(a));
        QuickSort.sort(a);
        System.out.println(Arrays.toString(a));
    }

    /**
     * 重新对数组元素进行排序
     * @param a
     */
    private static void shuffle(Comparable[] a) {
        int N = a.length;
        Comparable[] aux = new Comparable[N];
        Random random = new Random(21);
        for (int i=0; i<a.length; i++) {
            int k = i+random.nextInt(N-i);
            Comparable t = a[k];
            a[k] = a[i];
            a[i] = t;
        }
    }



    public static void sort(Comparable[] a) {
        shuffle(a);
        System.out.println(Arrays.toString(a));
        sort(a, 0, a.length-1);
    }

    private static void sort(Comparable[] a, int lo, int hi) {
        if (hi<=lo) return;
        int k = partition(a, lo, hi);
        sort(a, lo, k-1);
        sort(a, k+1, hi);
    }

    public static int partition(Comparable[] a, int lo, int hi) {
        int i=lo, j=hi+1;
        Comparable v = a[lo];
        while (true) {
            while (less(a[++i], v)) if (i==hi) break;
            while (less(v, a[--j])) if (j==lo) break;
            if (j<=i) break;
            exch(a, i, j);
        }
        exch(a, lo, j);
        return j;
    }

    private static void exch(Comparable[] a, int i, int j) {
        Comparable t = a[i];
        a[i] = a[j];
        a[j] = t;
    }

    private static boolean less(Comparable o1, Comparable o2) {
        return o1.compareTo(o2)<0;
    }
}
